src/rdkafka.h

```diff
@@ -3330,6 +3330,56 @@ int rd_kafka_queue_poll_callback (rd_kafka_queue_t *rkqu, int timeout_ms);
 
 
 /**
+ * @name Admin interface
+ *
+ */
+
+/**
+ * @brief Delete topic(s) in cluster.
+ *
+ * @param topics List of topic (names) to delete. Per-topic errors are
+ *        returned in the \c err field of each topic.
+ * @param timeout_ms Time for both controller lookup and actual topic deletion
+ *                   on the broker. If < 0 the absolute value is used
+ *                   for controller lookup timeout only and no timeout is
+ *                   used on the broker - the topic will only be marked for
+ *                   deletion. If > 0 the remaining time after controller
+ *                   lookup will be used as DeleteTopics request timeout
+ *                   on the broker, depicting how long the broker will
+ *                   wait for all replicas to have deleted the topic
+ *                   before responding to the client.
+ *                   Despite timeout setting, the topic will be marked
+ *                   for deletion.
```

Comment:
Specify that with timeout < 0, we receive an RD_KAFKA_RESP_ERR_REQUEST_TIMED_OUT for partition where topics where effectively marked as deletion, where we receive a different error if deletion is not possible (unknown topic, etc).

Reply:
I'm not into the idea of using negative numbers here - I'd have an additional parameter (probably a boolean, possibly an enum) to indicate the meaning of the timeout parameter.

Reply:
A "getCallback" boolean may be great indeed (with or without a C callback). And perhaps delete the `RD_KAFKA_RESP_ERR_REQUEST_TIMED_OUT` in case it is false then (as it is the response when there is no error)?

Reply:
@mhowlett The thing is that it is actually two timeouts baked into one:
 * controller lookup timeout, i.e., how long to wait for cluster metadata. This is mainly a problem for when this API is called right after client instantiation.
 * DeleteTopics protocol request which has its own **broker side** timeout (included in the request), telling the broker how long to wait for full delete propagation in the cluster before responding to the client.

From a user perspective though, there is typically only one timeout of concern: how long can my application wait for DeleteTopics to return, it is not really relevant if part of that time is spent looking up the controller.
And for power-users that understand that topic deletion is a best-effort operation time-wise, they might want to make sure the topic is marked for deletion but understand that there is no practical difference in waiting 100ms or 100seconds for the topic deletion to propagate, so they decide that the DeleteTopics protocol request should return immedatiately, only marking the topics for deletion, but they still want to allow X ms for looking up the controller - this is where the negative form comes in.
Yes, we could split that up into two parameters: lookup_timeout_ms and delete_timeout_ms, but I feel that is API bloat for something that will rarely be used.

Reply:
@treziac the positive timeout (or what remains of it after controller lookup) controls how long the request is blocked on the broker, if the topic has not been deleted by that time there will be no later propagation of the fact that the topics were deleted (or not), so a callback will not be much good.
Yes, we could use metadata, but that would require us to keep state between delete_topics() and future metadata calls, and there is still the risk that the topic is re-created between the delete and the next metadata query, in which case it will seem as if the topic was never deleted, but it was.

Reply:
I understand the use of negative timestamp, just that negative timestamp looks weird compared to only positive timestamp and bool saying if we wait for broker response during remaining timeout (and set request timeout to 0 or remaining accordingly as you do)
```
rd_kafka_admin_delete_topics (rd_kafka_t *rk,
                               rd_kafka_topic_partition_list_t *topics,
                               int timeout_ms, //always positive, negative will fail if controller lookup fail
                               bool use_remaining_timeout_for_delete_callback );
```
 but I may not be used on C way of dealing with time :) . 

I assume this mean we would only call it with negative value to return as soon as possible (with max timeout we want to give to controller lookup), or a kind of high (say min) if we want a "success" response assured from broker (and if no success, we will have to rely on metadata or check other  metrics like monitoring number of topics on broker or znode, but this is out of kafka client scope).

Here setting a non negative rather small timeout (eg  200ms) just mean "I marked topic for deletion in 20 ms, but wait some more time for response which will surely be a broker timeout")

It will be easy to map this on .NET anyway, so no issue here if it fits better like this in C.

Reply:
@edenhill - I think clarity/straightforwardness is more important to prioritize on than terseness of API. Maybe here you could have one overload with timeout_ms corresponding to the current implementation with positive timeout_ms, and another with both controller_timeout_ms and lookup_timeout_ms (as you suggest). I find this straightforward. I find the current setup requires mental energy to overcome the negative number weirdness.

Reply:
From a maintenance perspective it is imperative to keep the API surface at a bare minimum since these APIs must age well and be supported for the foreseeable future.
Adding multiple variations of the same API is not best practice in the C world, but rather a reluctant afterthought to fix an initial API design failure.
C users are typically accustomed to funky APIs that do require some documentation reading.

I do agree that this particular API is indeed somewhat funky with the negative timeout, encoding rich information in the signedness of a type, and from that perspective it might be better to have two timeouts: lookup_timeout_ms and request_timeout_ms, but does it end there? Should there be a network_timeout_ms and queue_timeout_ms as well? There are so many moving parts under the hood and we really just want to expose an API that is easy to use, but still somewhat flexible for hard-core users. And having developers understand and decide on proper timeouts has proven to be quite problematic, so the fewer things they need to be concerned with, the better.

In this case that means that for 99% of the users they just need to set a maximum timeout allowed for this call to do its thing, and if the return value is XYZ it means that topics were at least marked for deletion.
For the 1% power users, they can get the completely asynchronous request behaviour (instant timeout) without having to busy-loop until the controller is available (which is something we want to abstract from the user).

There are a couple of use-cases:
 1. lookup_timeout=X, request_timeout=0
 2. lookup_timeout=X, request_timeout=remaining(X)
 3. lookup_timeout=X, request_timeout=Y

The current scheme supports 1 and 2. If we believe there is a specific need for case 3 we'll need to add the second timeout as well, at the expense of an API that is harder to use for most users.

Reply:
I think the two method approach it's easier to comprehend and therefore easier to use and support, but this is just my 2c and API is ultimately up to you. In C# I will want to expose a less confusing API though, and method overloading is very natural there.

Reply:
We should probably go the Java way of providing an extensible Options type to Admin APIs.
E.g.;

```
 rd_kafka_admin_option_t *opt = rd_kafka_admin_option_new();
 rd_kafka_admin_option_set_timeout(opt, 1234); // block no longer than 1.234 seconds on the broker

 err = rd_kafka_admin_delete_topics(rk, myListOfTopics, opt, my_total_timeout_ms);
...
```